### ISSUE TITLE
Don't mark re-exported symbols as unused (F401)

### ISSUE THREAD

**Issue by layday** [charliermarsh/ruff#717]
Currently, an import in the form of:

```py
from a import b as b
```

will trigger F401, but this idiom is in widespread use as a type checker-friendly alternative to `__all__` which does not alter star import semantics.  Type checkers which support flagging unused imports will ignore re-exported symbols, i.e. symbols where the `as` target is identical to the original name:

```py
# pyright: strict

from os import environ  # Pyright reports this as unused
from os import environ as foo  # Likewise
from os import environb as environb   # No error
```

---

Thanks for ruff!

**Comment by charliermarsh** [comment]
This is easy to implement! My only question is whether it should be the default behavior, or designated under its own lint rule so as to be configurable (or via a configuration option).

Anyone have opinions on that? Maybe @andersk?

**Comment by charliermarsh** [comment]
I think this should be the default behavior since the re-export itself is effectively an explicit opt-in.

**Comment by andersk** [comment]
I think I agree. Although this was rejected upstream (PyCQA/pyflakes#474), it’s a pretty important pattern in the typed Python ecosystem (upstream is incorrect to expect that ecosystem to be entirely captured in PEPs). And I don’t see this significantly hurting anyone else.

**Comment by charliermarsh** [comment]
Going out in [v0.0.117](https://github.com/charliermarsh/ruff/releases/tag/v0.0.117) (building now).

**Comment by layday** [comment]
whoops, I should've probably mentioned the same holds for `import a as a`, although not as common.

**Comment by andersk** [comment]
Also apparently `import a.b.c as c`. (I find that one surprising, but mypy accepts it; [source](https://github.com/python/mypy/blob/v0.990/mypy/semanal.py#L2233).)

**Comment by charliermarsh** [comment]
Re-opening for those cases :)

**Comment by layday** [comment]
> Also apparently `import a.b.c as c`. (I find that one surprising, but mypy accepts it; [source](https://github.com/python/mypy/blob/v0.990/mypy/semanal.py?rgh-link-date=2022-11-13T22%3A03%3A04Z#L2233).)

Interesting, Pyright rejects that.

**Comment by andersk** [comment]
Yeah, Pyright seems more correct. Opened python/mypy#14086.
